### ISSUE TITLE
Support multiline matches in search: implement logical result and content retrieval parameter for search

### ISSUE THREAD

**Issue by rvantonder** [sourcegraph/sourcegraph#31028]
## Problem

Currently all content matches in search results conform to this type, the core assumption is that a matched range can only ever occur in one line:

```typescript
export interface ContentMatch {
  lineMatches: LineMatch[]
  ...
}

interface LineMatch {
  line: string // AKA preview
  lineNumber: number
  offsetAndLengths: number[][]
}
```

This representation is problematic for rendering or post processing data. Examples below, but I'm going to jump into the proposed solution. The point is that the time has come to fix this because it's actively blocking Code Insights/Compute and will also solve a host of longstanding and thorny issues. Most pertinent business/customer use case ("why are we doing this now?"): https://docs.google.com/document/d/1lNwfX-yGXx-sEcM7-X9iIAG6yFspJwBXXrzkPsYAuII/edit#heading=h.6mw6j1ignh29. A rough ask around timeline is to enable the multiline use case by end of April, if I recall correctly @Joelkw?

We'll need to coordinate effort here. @sourcegraph/search-core will be invaluable for helping identify/implement changes that we settle on, so I appreciate your eyes and thoughts here 🙏 . On the @sourcegraph/search-product side, myself and @camdencheek are in a good position to help with changing the result type on the frontend side and searcher side of things. We're also inviting @coury-clark to join the effort since this unlocks great things for Code Insights, and we'd like to enable him to make progress working on the search code base as far as Code Insights is concerned, so that this doesn't feel blocked on just search team progress/availability.

## Proposed solution

We need to fix two parts:

**(1) Represent the content matched as a single, logical result irrespective of how many lines it has**

```go
type FileMatch struct {
  ...
  matches Range[]
  ...
}

type Range struct {
  Start Location 
  End   Location 
}

type Location struct {
  Line int
  Column int
}
```

To do this, we need to update searcher and Zoekt to do bookkeeping of ranges. For single-line matches, there's an isomorphism from `LineMatches` to this representation, so we already have code to do _some_ bookkeeping. I don't recall the internals well enough to estimate how much effort this is. Having matches represented this way will stop regressions with match counts, make rendering logic easier, and also enable linking to ranges of associated matches in our UI/webview from Code Insights or compute applications.

**(2) We need to allow clients to optionally request the content associated with a logical result.**

We benefit from being able to request content for a result in the following cases:

- Compute / Code Insights can work with entirety of content to efficiently calculate capture groups (alternatively, it would need to request all file content from gitserver)
- `src-cli` can effectively render logical matches based on matched content and avoid fetching file content from gitserver.

In these cases, the requested content will replace and supersede the role of `preview` in our current result type.

Note that we _don't_ want to include the `preview` or result `content` by default in the main Sourcegraph app. This because the typical code path has the client getting search results from the backend, then request the highlighted file (i.e., "content") from the backend, and then only uses the `LineMatch` ranges to compute the highlighting/match overlay. We want to continue with this architecture, and only respond to our main app's client with match ranges and migrate away from `LineMatches`. We do not need to send it the actual content. There is a separate effort around streaming highlighted content, but that's an orthogonal issue to this issue of line-based results versus logical results (our streaming highlighting effort _also_ relies on the line-based result type currently, which is not good).

So the point is, we likely need a request parameter to our search backends (searcher, Zoekt) to optionally return result content.

---

An important for Code Insights is to efficiently compute capture groups for (potentially multiline) search results. After discussion, I believe the easiest and cleanest path to get there is to implement the above by first tackling part (2) which means:

- adding a parameter to our search backend API (Zoekt and searcher) to request it hydrate a new `content` field in `FileMatch` result type. Over time, we deprecate and remove `preview` and `content` supersedes `preview`. This is the direction we've wanted to go in for a while, so it aligns with "better state of search code". We cannot deprecate `preview` right away for `<reasons>`. Thoughts here @sourcegraph/search-core, specifically for Zoekt?

- capturing the content for a logical match in our backends. Here I don't remember the details for searcher and Zoekt and how much effort it is to do this. Please chime in @sourcegraph/search-core if you have an intuition.

We can complete this part (2) without doing part (1), or migrating away from `LineMatch` just yet. But while we're doing (2), it's useful to think about how we could add the bookkeeping to our backends to capture the full range of matches. Over time, but as a less urgent case, we want to represent matches as a list of these absolute ranges to overcome all the snags we have with the current line matches, and to unlock things like https://github.com/sourcegraph/sourcegraph/issues/30711 later.

Alternatives / related thoughts: It's possible we could add functionality to our search backends to natively recognize capture groups, but I honestly don't know how big this ask is, or whether it makes sense architecturally yet. Doing something like that will need a larger change to the result type coming from our backends, and I don't think that footprint is worth it right now. As a matter of our current needs, empirically Code Insights / Compute can operate efficiently if it can get the content to process capture groups and I don't think there's a strong need to add native capture group support to backends.

## Next steps

Mainly open questions for @sourcegraph/search-core (@keegancsmith @ggilmore): does this direction make sense to you? Ideas around effort and availability on the core team to help with this (esp. for modification to Zoekt?)

@camdencheek and @coury-clark feel free to discuss/comment if the direction makes sense to you. I know @camdencheek could probably estimate some effort for doing the plumbing in frontend/searcher. From there we can do a walkthrough of the code with Coury and work together on changes.

---

More context, why a line-based result type is not great for our tooling.

Even though our search backend (searcher and Zoekt) can find matches for multiline regex patterns like `foo\nbar`, multiline content that match the pattern are broken up to conform to the line-based assumption the above type. When multiline content matches a pattern, and we jam it into this line-based result type, we throw away the information that tells us which lines correspond logically to the pattern. In a read-only client, where we simply render highlighted ranges (potentially across multilines), we get away with the line-based assumption, since the user can likely visually pick out how the pattern corresponds to highlighted lines. But clients don't actually know which line ranges a matching pattern corresponds to, we've thrown that information away by the time it gets to the client. This one assumption has been a cause of many troubles:

- inaccurate result counts. Here we would count the number of lines as a single result, since our type only tracks lines, and we have no way to tell how many of them correspond to the "logical" match (i.e., the one that corresponds to the pattern. https://github.com/sourcegraph/sourcegraph/issues/9426 

- inconsistent expectations from the search API. Using the line-based assumption, clients may assume that complete information about the content should be exposed in the `Preview` or `line` field, and use that content data for rendering. These assumptions break down when using structural search which inherently does not consider a single line as any kind of match or result--a match is always qualified absolutely as a range over lines and columns (or, alternatively, absolute offsets in a file). https://github.com/sourcegraph/sourcegraph/issues/11134

- The line-based content of `Preview`/`line` stops us from building data processing on top of results. When a process wants to extract capture groups (or any similar process that needs to match over the _entirety_ of some logical result), we can't rely on any search result like `Preview` or `LineMatch` offsets since it can't tell us what the extent of that entire result is. We'd have to process an entire file to get the result. This last case is the straw that breaks the camel's back that now pushes us to finally address the shortcomings of the line-based assumption.

**Comment by keegancsmith** [comment]
This sounds good to me. General context: searching in all cases does extra work to hydrate in the line preview. When we do a search all we actually have are the document and offsets, we then do extra work for the benefit of the client. In fact even calculating what line number we are on is extra work that we do. Our clients don't actually use preview lines so this sounds like good stuff to me.

Do you need the line numbers if we are returning contents? Surely byte offsets are more useful to you? They also don't suffer from 1-based vs 0-based indexing inconsistencies. What _exactly_ do you think is needed in sourcegraph. I would imagine what we find useful are:

- file contents for syntax highlighting and avoiding speaking to gitserver
- byte offsets of what to highlight in the file
- byte offsets of the line?
- Maybe byte offsets for contextual lines like 2 lines before/after? This is likely less useful since we do work to combine displayed hunks in our display.

Zoekt:
- Zoekt already has an API option to return file contents. I believe structural search uses this. Our streaming API will need to support this if it doesn't already.
- Changing the zoekt API to match what you describe sounds good to me. We need to keep the old API somehow since for example other clients (like zoekt-webserver HTML interface) use the line preview. We are aware of other zoekt users outside of sourcegraph who use the line preview.

**Comment by Joelkw** [comment]
> A rough ask around timeline is to enable the multiline use case by end of April, if I recall correctly @Joelkw?

Correct! It'd be amazing if it landed in the April 20 release, but I assume more realistic is the May release (so fully end of April).

**Comment by camdencheek** [comment]
> Do you need the line numbers if we are returning contents? Surely byte offsets are more useful to you? They also don't suffer from 1-based vs 0-based indexing inconsistencies. What exactly do you think is needed in sourcegraph.

This has also been my biggest set of open questions. 

> file contents for syntax highlighting and avoiding speaking to gitserver

I think this would be ideal from a "maximum flexibility" perspective, but my problem with this approach is we throw away a massive number of results. We throw away results when: 1) display limit is hit, so we only count the results, 2) `select:` is used, so we don't care about the file contents, 3) we combine results in `and/or` queries, 4) limit is hit, so we throw away the remainder of a batch. In all of these cases, especially display limit, I don't think we want to be paying the overhead of sending the full file contents from zoekt/searcher/gitserver. 

> byte offsets of what to highlight in the file

I think we want at least this. Though I think there is an open question of whether we should be using byte offsets or rune offsets. Byte offsets let us just index the file content (useful if we have the contents available and want to get the match content), but rune offsets are easier to use once we get to TypeScript land (we depend on rune offsets for displaying the line highlights). The example Rijnard gave uses line/column for `Location`, but I think we want byte/rune offsets as well for easy indexing. 

> byte offsets of the line?

If we have the file contents, I don't think so. With line byte offsets for lines, we could easily index the file content for the start/end of line so we can get the line that a match is on. However, we can also (relatively) efficiently find the first newline before and after a match if we have the file contents, so I don't think this is needed. 

> Maybe byte offsets for contextual lines like 2 lines before/after? This is likely less useful since we do work to combine displayed hunks in our display.

Agree that this is less useful. Additionally, I'd rather avoid any sort of non-general solution here. If we want 1 line of context somewhere, we'll probably also want 2 lines further down the line. Having the file content available also makes this easy to get.

**Comment by camdencheek** [comment]
I would really love to just _have_ file contents available on the `FileMatch` type, but that could significantly increase our network usage. Especially since we pretty much only use it for the first ~50 files.

**Comment by camdencheek** [comment]
Realistically, I think the most reasonable solution might just to be to hydrate the file contents lazily like we do repo stats. I don't think we actually _need_ the contents until then, even if we do highlight-before-stream. Then we would change the match range type to something like the following in our backend. We'd need to adjust everything to use byte-indexing rather than rune-indexing, but this should be significantly more efficient than rune-scanning the entire file content every time we want to index into it. 

```
type Range struct {
  Start Location 
  End   Location 
}

type Location struct {
  // byte offsets
  Start int
  End int
  
  // line/column for convenience? This is most efficient to do in the backends, but we can also get this when we actually need it. Since we are likely only doing this for a few files, it doesn't seem unreasonable to me to spend the CPU cycles on generating line/column as needed. 
  Line int
  Column int
}
```

**Comment by camdencheek** [comment]
> @camdencheek could probably estimate some effort for doing the plumbing in frontend/searcher. From there we can do a walkthrough of the code with Coury and work together on changes.

I spiked this out a few days ago and got searcher to start using `Range` types like defined above. However, the biggest question here is "what do we do about transmitting content." Once we decide on how this is going to work, I expect this to only take a couple of days of focused work.

**Comment by keegancsmith** [comment]
My answers below are gonna be very zoekt specific. But maybe you can take these ideas and generalise? eg content being optional like in Zoekt. Also display limits and stats.

> > file contents for syntax highlighting and avoiding speaking to gitserver
> 
> I think this would be ideal from a "maximum flexibility" perspective, but my problem with this approach is we throw away a massive number of results. We throw away results when: 1) display limit is hit, so we only count the results

Zoekt has display limit. We can then rely on its Stats object for counting instead of doing it ourselves.

> 2) `select:` is used, so we don't care about the file contents

We don't have to always ask for contents. It can be optional. EG don't ask when doing select. Also we should tell zoekt we are doing select so it doesn't do extra work anyways.

> 3) we combine results in `and/or` queries

Zoekt should do and/or. Harder to say for our other backends? but we should probably start thinking about that.

> 4) limit is hit, so we throw away the remainder of a batch.

When we hit limit we cancel the context though right so we other than things sitting in network buffers it should all be good?

> > byte offsets of what to highlight in the file
> 
> I think we want at least this. Though I think there is an open question of whether we should be using byte offsets or rune offsets. Byte offsets let us just index the file content (useful if we have the contents available and want to get the match content), but rune offsets are easier to use once we get to TypeScript land (we depend on rune offsets for displaying the line highlights). The example Rijnard gave uses line/column for `Location`, but I think we want byte/rune offsets as well for easy indexing.

It is easier to go from byte offsets to rune offsets than the other way around. Especially if we have file content. Also the rune offsets are just on the line, so that is even easier to do :)

> > byte offsets of the line?
> 
> If we have the file contents, I don't think so. With line byte offsets for lines, we could easily index the file content for the start/end of line so we can get the line that a match is on. However, we can also (relatively) efficiently find the first newline before and after a match if we have the file contents, so I don't think this is needed.

Agreed. If we are sending the line (or file) down to the end user, this is trivial to calculate vs that / with that.

> I would really love to just _have_ file contents available on the `FileMatch` type, but that could significantly increase our network usage. Especially since we pretty much only use it for the first ~50 files.

I think we should start more heavily leaning on Stats from backends so we can transfer less and rely more on our backends telling us stats past the display limits.

> Realistically, I think the most reasonable solution might just to be to hydrate the file contents lazily like we do repo stats. I don't think we actually _need_ the contents until then, even if we do highlight-before-stream. Then we would change the match range type to something like the following in our backend. We'd need to adjust everything to use byte-indexing rather than rune-indexing, but this should be significantly more efficient than rune-scanning the entire file content every time we want to index into it.

Indeed, this is realistic and the most incremental.
>   // line/column for convenience? This is most efficient to do in the backends, but we can also get this when we actually need it. Since we are likely only doing this for a few files, it doesn't seem unreasonable to me to spend the CPU cycles on generating line/column as needed. 

I think keeping it simple and just using byte offsets makes sense to me. If every file we actually send to the user requires hydration from gitserver/etc then we can calculate line numbers then?

> I don't know that this solves the codeinsights use-case though, since they will need to fetch the contents for every result in order to get the match contents.

Maybe this points back to making filling in these values optional? IE whatever does the searching anyways is doing this. The proposed change just makes us avoid it most of the time.

**Comment by benvenker** [comment]
converting to ZenHub epic for ease of (my) tracking

**Comment by camdencheek** [comment]
For watchers of this issue, I plan to start on this in the next couple of weeks. This is slotted on into our foundational technical improvements slated for Q2.

**Comment by rvantonder** [comment]
Closing this as "backend work is complete". Follow up issues for client side:

https://github.com/sourcegraph/sourcegraph/issues/38891
https://github.com/sourcegraph/sourcegraph/issues/11134
